{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>6mA motif identification and methylome characterization - tutorial</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This workflow demonstrates how to use tools I created for the identification and analysis of RM target motifs from Oxford Nanopore sequencing reads. The workflow starts with the BED file generated by ONT's ModKit program, and is broken down into 3 modules: \n",
    "\n",
    "<b>1. Motif Discovery </b><br>\n",
    "Identifies target motifs of the RM system(s) harbored in your genome and calculates the percentage of sites within the motif genome wide that are modified. <p>\n",
    "<p>\n",
    "<b>2. Methylome Profiler </b> <br>\n",
    "Maps all modified bases to your genome, counts modified bases within all coding and intergenic regions, counts across all genomic elements, sliding window methylation rates<p>\n",
    "<b>3. Motif Analysis </b><br>\n",
    "Statistical testing on whether identified RM target motifs are enriched or avoided in different areas of your genome <p>\n",
    "\n",
    "To get started, clone the github repo to your local machine:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clone github repo \n",
    "! git clone https://github.com/bely12/borrelia-methylomics.git"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>MODULE 1: MOTIF DISCOVERY</h3>\n",
    "<h4><b>Get sequences surrounding all modified adenines of length <i>k</i></b></h4>\n",
    "<b>Script:</b> Akmaker.py <p>\n",
    "\n",
    "<b>Inputs:</b> BED file containing modified base call data and the reference genome that was used to map your reads <p>\n",
    "\n",
    "<b>Output:</b> fasta of all 6mA centered kmers<p>\n",
    "\n",
    "Use <i>-len</i> argument to specify the size of the kmer, as well as <i>-depth</i> for miniumum sequencing depth to consider (default >10) and <i>-mod_thresh</i> to set a frequency threshold for calling a base modified (default >50). The -modified option is used in this example. Use --help for more output options. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract kmers surrounding all modified adenine positions as a multifasta file\n",
    "! python3 Motif_Discovery/Akmaker.py \\\n",
    "    -bed test_files/test.bedmethyl \\\n",
    "    -ref test_files/test.fasta \\\n",
    "    -modified \\\n",
    "    -len 11 \\\n",
    "    -out test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b><h4>Find most commonly modfied short sequences containing your modified adenine</h4></b>\n",
    "\n",
    "<b>Script:</b> motif_discovery.py <p>\n",
    "\n",
    "<b>Inputs:</b> BED file, reference genome, set of all 6mA centered kmers (Akmaker.py output) <p>\n",
    "\n",
    "<b>Output:</b> List of most abundant short sequences, total number of occurences, and modifification frequency genome wide <p>\n",
    "\n",
    "Specify the 0-indexed position of the 6mA with <i>-mod_pos_index</i>. The <i>-target_lengths</i> argument specifies the motif lengths to consider (separate different numbers with a comma). Use <i>--help</i> to customize other parameters such as miniumum sequencing depths and mod call thresholds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "motif \t occurences \t mod \t percent\n",
      "ATAAA \t 7 \t 6 \t 0.857\n",
      "GATAA \t 15 \t 12 \t 0.8\n",
      "CGATA \t 30 \t 30 \t 1.0\n",
      "CGGTA \t 15 \t 15 \t 1.0\n",
      "CGAGA \t 17 \t 15 \t 0.882\n",
      "GATAAA \t 7 \t 6 \t 0.857\n",
      "CGATAA \t 13 \t 11 \t 0.846\n",
      "CGAGAA \t 10 \t 10 \t 1.0\n",
      "TCGATA \t 11 \t 11 \t 1.0\n",
      "CGATAT \t 9 \t 9 \t 1.0\n",
      "GATAAAT \t 3 \t 2 \t 0.667\n"
     ]
    }
   ],
   "source": [
    "! python3 Motif_Discovery/motif_discovery.py \\\n",
    "    -bed test_files/test.bedmethyl \\\n",
    "    -ref test_files/test.fasta \\\n",
    "    -candidates test_positive_seqs.fasta \\\n",
    "    -mod_pos_index 5 \\\n",
    "    -target_lengths 5,6,7 \\\n",
    "    -n_targets 5 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b><h4>Fine tune short sequences to a definitive R-M target motif</h4></b>\n",
    "\n",
    "<b>Script:</b> mod_calc.py <p>\n",
    "\n",
    "<b>Inputs:</b> motif or list of motifs and the position of the modified adenine <p>\n",
    "\n",
    "<b>Output:</b> Total number of occurences, total number modified, and modifification frequency genome wide for each specified motif <p>\n",
    "\n",
    "Based on results from motif_discovery.py, look up modification frequencies inferred motif(s). Motifs can be entered with <i>-motif</i> with ambiguous nucleotide symbols. Use 1-based index for specifiying the position of the modified adenine in your motif with <i>-mod_pos</i>. Use <i>--help</i> to see options for custom parameters/thresholds. Run multiple times, substituting N's at different positions to hammer down exact specificity of the R-M target and determine a consensus motif. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "motif \t occurences \t mod \t percent\n",
      "CGAGA \t 17 \t 15 \t 0.882\n",
      "CGATA \t 30 \t 30 \t 1.0\n",
      "CGGGA \t 12 \t 12 \t 1.0\n",
      "CGGTA \t 15 \t 15 \t 1.0\n",
      "\n",
      "\n",
      "found in genome =  74\n",
      "modified in genome =  72\n",
      "Total percent modfied for all motifs =  97.3\n"
     ]
    }
   ],
   "source": [
    "! python3 Motif_Discovery/mod_calc.py \\\n",
    "    -bed test_files/test.bedmethyl \\\n",
    "    -ref test_files/test.fasta \\\n",
    "    -motif CGRKA \\\n",
    "    -mod_pos 5 \\ "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>MODULE 2: METHYLOME PROFILER</h3>\n",
    "<h4><b>Generate a table of all modified base positions genome wide</b></h4>\n",
    "\n",
    "<b>Script:</b> mod_mapper.py <p>\n",
    "\n",
    "<b>Inputs:</b> BED file, reference genome, modified motif, or file containing a list of multiple modified motifs and the position of the modified adenine <p>\n",
    "\n",
    "<b>Output:</b> table containing coordinates of all modified target and off-target sites, as well as target sites that are not modified<p>\n",
    "\n",
    "Use <i>--help</i> to customize other parameters such as miniumum sequencing depth to include and mod call thresholds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python3 Methylome_Profiler//mod_mapper.py \\\n",
    "    -bed test_files/test.bedmethyl \\\n",
    "    -ref test_files/test.fasta \\\n",
    "    -motif CGRKA \\\n",
    "    -mod_pos 5 \\\n",
    "    > test.sites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View sites table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif</th>\n",
       "      <th>plasmid_id</th>\n",
       "      <th>position</th>\n",
       "      <th>strand</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>off_target</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>564</td>\n",
       "      <td>-</td>\n",
       "      <td>modified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>off_target</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>565</td>\n",
       "      <td>-</td>\n",
       "      <td>modified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGAGA</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>566</td>\n",
       "      <td>-</td>\n",
       "      <td>modified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGAGA</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>1630</td>\n",
       "      <td>-</td>\n",
       "      <td>unmodified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGATA</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>2848</td>\n",
       "      <td>+</td>\n",
       "      <td>modified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         motif     plasmid_id  position strand       status\n",
       "0  off_target    NC_001849.2        564     -      modified\n",
       "1  off_target    NC_001849.2        565     -      modified\n",
       "2       CGAGA    NC_001849.2        566     -      modified\n",
       "3       CGAGA    NC_001849.2       1630     -    unmodified\n",
       "4       CGATA    NC_001849.2       2848     +      modified"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "sites_file = 'test.sites'\n",
    "sites=pd.read_table(sites_file, header=None)\n",
    "sites = sites.rename(columns={0: 'motif', 1: 'plasmid_id', 2: 'position', 3: 'strand', 4: 'status'})\n",
    "sites.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b><h4>Link modified sites to features in the genome</h4></b>\n",
    "\n",
    "<b>Script:</b> annotate_methylome.py <p>\n",
    "\n",
    "<b>Inputs:</b> .gbff file and modified sites table  <p>\n",
    "\n",
    "<b>Output:</b><br>\n",
    "3 tsv files named as below following user assigned prefix\n",
    "mod_annotations.tsv: table of modified base counts and normalized rates within all cds and intergenic regions of genome <br>\n",
    "annotated_sites.tsv: table of modified sites with column for the gene ID corresponding to the coordinate (or interegenic if not within cds) <br>\n",
    "summary.tsv: table of modified base counts and normalized rates for each chromosome/plasmid in the genome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python3 Methylome_Profiler/annotate_methylome.py \\\n",
    "    -gbff test_files/B31.gbff \\\n",
    "    -sites test.sites \\\n",
    "    -out test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View feature counts table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>element_id</th>\n",
       "      <th>len</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>feature_len</th>\n",
       "      <th>strand</th>\n",
       "      <th>target_mod</th>\n",
       "      <th>mod_rate</th>\n",
       "      <th>target_no_mod</th>\n",
       "      <th>unmodified_rate</th>\n",
       "      <th>off_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1457</th>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>16821</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>na</td>\n",
       "      <td>1</td>\n",
       "      <td>328</td>\n",
       "      <td>328</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>16821</td>\n",
       "      <td>BB_D01</td>\n",
       "      <td>WP_010890249.1</td>\n",
       "      <td>329</td>\n",
       "      <td>803</td>\n",
       "      <td>475</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>16821</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>na</td>\n",
       "      <td>804</td>\n",
       "      <td>869</td>\n",
       "      <td>66</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1460</th>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>16821</td>\n",
       "      <td>BB_RS05615</td>\n",
       "      <td>WP_012614969.1</td>\n",
       "      <td>870</td>\n",
       "      <td>1020</td>\n",
       "      <td>151</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>16821</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>na</td>\n",
       "      <td>1021</td>\n",
       "      <td>1408</td>\n",
       "      <td>388</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       element_id    len     gene_id      protein_id  start   end  \\\n",
       "1457  NC_001849.2  16821  intergenic              na      1   328   \n",
       "1458  NC_001849.2  16821      BB_D01  WP_010890249.1    329   803   \n",
       "1459  NC_001849.2  16821  intergenic              na    804   869   \n",
       "1460  NC_001849.2  16821  BB_RS05615  WP_012614969.1    870  1020   \n",
       "1461  NC_001849.2  16821  intergenic              na   1021  1408   \n",
       "\n",
       "      feature_len strand  target_mod  mod_rate  target_no_mod  \\\n",
       "1457          328     na           0      0.00              0   \n",
       "1458          475      +           1      2.11              0   \n",
       "1459           66     na           0      0.00              0   \n",
       "1460          151      +           0      0.00              0   \n",
       "1461          388     na           0      0.00              0   \n",
       "\n",
       "      unmodified_rate  off_target  \n",
       "1457              0.0           0  \n",
       "1458              0.0           2  \n",
       "1459              0.0           0  \n",
       "1460              0.0           0  \n",
       "1461              0.0           0  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read file into pandas df\n",
    "feature_counts_file = 'test_mod_annotations.tsv'\n",
    "feature_counts=pd.read_table(feature_counts_file)\n",
    "#filter for element_id's within test files \n",
    "feature_counts = feature_counts[ ((feature_counts['element_id'] == 'NC_001849.2') | (feature_counts['element_id'] == 'NC_001852.1'))]\n",
    "\n",
    "feature_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View site annotation table "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>motif</th>\n",
       "      <th>plasmid_id</th>\n",
       "      <th>position</th>\n",
       "      <th>strand</th>\n",
       "      <th>status</th>\n",
       "      <th>annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>off_target</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>564</td>\n",
       "      <td>-</td>\n",
       "      <td>modified</td>\n",
       "      <td>BB_D01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>off_target</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>565</td>\n",
       "      <td>-</td>\n",
       "      <td>modified</td>\n",
       "      <td>BB_D01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGAGA</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>566</td>\n",
       "      <td>-</td>\n",
       "      <td>modified</td>\n",
       "      <td>BB_D01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGAGA</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>1630</td>\n",
       "      <td>-</td>\n",
       "      <td>unmodified</td>\n",
       "      <td>BB_D04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGATA</td>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>2848</td>\n",
       "      <td>+</td>\n",
       "      <td>modified</td>\n",
       "      <td>intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        motif   plasmid_id  position strand      status       annot\n",
       "0  off_target  NC_001849.2       564      -    modified      BB_D01\n",
       "1  off_target  NC_001849.2       565      -    modified      BB_D01\n",
       "2       CGAGA  NC_001849.2       566      -    modified      BB_D01\n",
       "3       CGAGA  NC_001849.2      1630      -  unmodified      BB_D04\n",
       "4       CGATA  NC_001849.2      2848      +    modified  intergenic"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_file = 'test_annotated_sites.tsv'\n",
    "annot_sites=pd.read_table(annot_file)\n",
    "annot_sites.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "View methylation summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>len</th>\n",
       "      <th>total_mods</th>\n",
       "      <th>total_rate</th>\n",
       "      <th>cds_mods</th>\n",
       "      <th>cds_rate</th>\n",
       "      <th>intergenic_mods</th>\n",
       "      <th>intergenic_rate</th>\n",
       "      <th>off_target_mods</th>\n",
       "      <th>off_target_rate</th>\n",
       "      <th>unmodified_targets</th>\n",
       "      <th>unmodified_target_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001849.2</td>\n",
       "      <td>16821</td>\n",
       "      <td>18</td>\n",
       "      <td>1.070</td>\n",
       "      <td>7</td>\n",
       "      <td>0.879</td>\n",
       "      <td>11</td>\n",
       "      <td>1.240</td>\n",
       "      <td>10</td>\n",
       "      <td>0.594</td>\n",
       "      <td>1</td>\n",
       "      <td>0.059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_001852.1</td>\n",
       "      <td>29766</td>\n",
       "      <td>54</td>\n",
       "      <td>1.814</td>\n",
       "      <td>53</td>\n",
       "      <td>1.932</td>\n",
       "      <td>1</td>\n",
       "      <td>0.417</td>\n",
       "      <td>27</td>\n",
       "      <td>0.907</td>\n",
       "      <td>1</td>\n",
       "      <td>0.034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id    len  total_mods  total_rate  cds_mods  cds_rate  \\\n",
       "3  NC_001849.2  16821          18       1.070         7     0.879   \n",
       "8  NC_001852.1  29766          54       1.814        53     1.932   \n",
       "\n",
       "   intergenic_mods  intergenic_rate  off_target_mods  off_target_rate  \\\n",
       "3               11            1.240               10            0.594   \n",
       "8                1            0.417               27            0.907   \n",
       "\n",
       "   unmodified_targets  unmodified_target_rate  \n",
       "3                   1                   0.059  \n",
       "8                   1                   0.034  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_file = 'test_mod_summary.tsv'\n",
    "summary_table=pd.read_table(summary_file)\n",
    "summary_table = summary_table[ ((summary_table['id'] == 'NC_001849.2') | (summary_table['id'] == 'NC_001852.1'))]\n",
    "summary_table.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
