{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Workflow for processing raw nanopore sequencing output</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Step 1 – Basecalling on raw pod5 files (if basecalling was not performed in real time)</h3>\n",
    "\n",
    "<p>\n",
    "<em>\n",
    "Requires Dorado downloaded and installed; see\n",
    "<a href=\"https://github.com/nanoporetech/dorado\">https://github.com/nanoporetech/dorado</a>\n",
    "for instructions and help\n",
    "</em>\n",
    "</p>\n",
    "Here you are performing basecalling on your raw nanopore signal data (stored in pod5 files) and likely aligning the reads to a reference genome. The output is a bam file. If you also want a fastq of the reads, add --emit-fastq as an argument<p>\n",
    "\n",
    "<p>When selecting your basecalling model consider:</p>\n",
    "\n",
    "<ul>\n",
    "  <li><strong>Speed / accuracy types:</strong> fast | hac | sup</li>\n",
    "  <li><strong>Type of bases:</strong> canonical | modified</li>\n",
    "</ul>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look up basecalling models if you need to\n",
    "! dorado --list-models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# basecalling\n",
    "! dorado basecaller [basecalling model id] [path to pod5] --reference [path to reference genome] > output.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Step 2 – Sort and index BAM file, get coverage statistics</h3>\n",
    "<p>\n",
    "<em>\n",
    "Requires samtools; see:\n",
    "<a href=\"https://www.htslib.org/doc/\">https://www.htslib.org/doc/</a>\n",
    "for instructions and help\n",
    "</em>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort\n",
    "! samtools sort -o [output.sorted.bam] [output.bam]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# index\n",
    "! samtools index [output.sorted.bam]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coverage statistics\n",
    "! samtools coverage [output.sorted.bam]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Step 3 – Get counts for modified base positions (if you used modified basecalling)</h3>\n",
    "\n",
    "<p>\n",
    "<em>\n",
    "Requires Modkit installed; see\n",
    "<a href=\"https://github.com/nanoporetech/modkit\">https://github.com/nanoporetech/modkit</a>\n",
    "for instructions and help\n",
    "</em>\n",
    "</p>\n",
    "This aggregates reads on your modified base and counts the number of modified and unmodified calls per position. Output is a bed file with methylation summaries.<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# modified base data\n",
    "! modkit pileup [output.sorted.bam] [output.bedmethyl]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>Other useful commands</h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Aligning reads that have already been basecalled</h3>\n",
    "<p>\n",
    "Do this for unaligned fastq or bam files; dorado uses minimap2 for aligning reads to reference genome\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align reads to reference\n",
    "! dorado aligner [reference genome] [reads.fastq] > aligned.bam "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Demultiplexing pooled barcode reads</h3>\n",
    "<p>\n",
    "If you didn't do real time basecalling during sequencing, you will have to do it prior to demultiplexing. After basecalling (see Step 1), use this command to sort your fastq basecalled reads by sample barcode. The output is a bam file for each barcode. It will be written to a directory you specifiy (or create a new directory if a desired one doesn't exist).\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# demultiplexing reads\n",
    "! dorado demux [reads.fastq] --kit-name [sequencing kit code] --output-dir [/demux_reads]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Grab list of read id's for a specific barcode</h3>\n",
    "<p>\n",
    "Do this if you need to subset reads from a pod5 that contains all barcoded samples. This grabs read id's from a bam file containing reads for a specific barcode. \n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab read id by barcode\n",
    "! samtools view [barcode07.bam] | cut -f1 | sort -u > read_ids.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Subset a pod5 file to reads for a specific barcode</h3>\n",
    "<p>\n",
    "<em>\n",
    "Requires pod5 package installed; see\n",
    "<a href=\"https://software-docs.nanoporetech.com/pod5/latest/tools/\">https://software-docs.nanoporetech.com/pod5/latest/tools/</a>\n",
    "for instructions and help\n",
    "</em>\n",
    "<p>\n",
    "Do this to generate a new pod5 file with only a barcodes for a specific sample. Helpful if basecalling needs to be done separately for a specific barcoded sample. You will need your list of read id's, one per line (see above command). Name of output file must end in the .pod5 extension for dorado to recognize it later. Use --missing-ok arg to specificy that its ok that not every original pod5 file will have read id's for your barcode, otherwise the command will fail. --output can also be a new directory if you want to separate it (usually a good idea since you point dorado to directories of pod5 files to execute commands)\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset pod5\n",
    "! pod5 filter [pod5_directory/] --ids [read_ids.txt] --output [new_pod5_directory/subset.pod5] --missing-ok"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Generate a consensus sequence for a targeted region of genome from reads that are aligned to a reference genome</h3>\n",
    "<p>\n",
    "This workflow shows an example for generating a consensus sequence for an RM gene in the B331 strain found on lp25 with the coordinates 423-4238. The fasta id for lp25 on the reference genome is B331_CP017212_116_lp25. \n",
    "</p>\n",
    "Step 1 - subset bam file for only reads in the region of interest and generate a new consensus sequence\n",
    "<ul>\n",
    "  <li>Exact id from reference fasta needs to be used, followed by \":start-end\"</li>\n",
    "  <li>subset > sort > index > make region specific ref seq > check number of aligned reads</li>\n",
    "</ul>\n",
    "<p>\n",
    "<p>\n",
    "Step 2 - Call variantws and get consensus sequence\n",
    "<ul>\n",
    "   <li>variant call > zip vcf file > index vcf file > generate consensus sequence</li>\n",
    "   <li>*Note -Ov specifies plain VCF output rather than compressed; I think the newest version of bcftools can take uncompressed as input for consensus, but I ended up needing to compress</li>\n",
    "</ul>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### step 1 samtools workflow ###\n",
    "# subset for reads overlapping region of interest\n",
    "samtools view -b B331.sorted.bam B331_CP017212_116_lp25:423-4238 > gene_reads.bam\n",
    "\n",
    "# sort\n",
    "samtools sort -o gene_reads.sorted.bam gene_reads.bam\n",
    "\n",
    "# index\n",
    "samtools index gene_reads.sorted.bam\n",
    "\n",
    "# create reference seq for region of interest\n",
    "samtools faidx B331.fasta B331_CP017212_116_lp25:423-4238 > gene_ref.fasta\n",
    "\n",
    "# if you want to know how many reads in the new bam\n",
    "samtools view -c gene_reads.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### step 2 vcftools workflow ###\n",
    "# call variants\n",
    "bcftools mpileup -f B331.fasta gene_reads.sorted.bam -Ov -o gene.vcf\n",
    "\n",
    "# zip \n",
    "bgzip gene.vcf\n",
    "\n",
    "# index\n",
    "bcftools index gene.vcf.gz\n",
    "\n",
    "# get consensus seqeunce\n",
    "bcftools consensus -f gene_ref.fasta gene.vcf.gz > consensus.fasta"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
