{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cc720c06",
   "metadata": {},
   "source": [
    "<H2>Generate seqs for building a species tree via aligning reads or assemblies to single ref genome then calling variants</H2>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0ec5b17",
   "metadata": {},
   "source": [
    "<H3>I. Generating consensus seqs from fastq sequencing reads</H3>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a94bb96c",
   "metadata": {},
   "source": [
    "<B>Step 1 - Map fastq reads from ONT output to reference genome using minimap2</B>\n",
    "\n",
    "inputs: reference fasta, sequencing reads in fastq format<br>\n",
    "output: sam file of aligned reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "864e57b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "minimap2 -ax map-ont ref.fasta reads.fastq > strain.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0106f3a5",
   "metadata": {},
   "source": [
    "<B>Step 2 - Convert SAM to BAM, sort, index</B>\n",
    "\n",
    "inputs: SAM of aligned reads from minimap2<br>\n",
    "output: sorted and indexed BAM file of aligned reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e027d75b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sam to bam\n",
    "samtools view -bS strain.sam > strain.bam\n",
    "\n",
    "# sort\n",
    "samtools sort -o strain.sorted.bam strain.bam\n",
    "\n",
    "#index\n",
    "samtools index strain.sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad85872a",
   "metadata": {},
   "source": [
    "<B>Step 3 - BCF tools mpileup to get genotype likihoods</B>\n",
    "\n",
    "inputs: reference fasta, sorted bam<br>\n",
    "output: bcf pileup file<br>\n",
    "*notes: -Ou specifies uncompressed output (otherwise I believe output is compressed by default?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2422aa22",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools mpileup -f ref.fasta strain.sorted.bam -Ou -o strain.mpileup.bcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948dcc13",
   "metadata": {},
   "source": [
    "<B>Step 4 - Call variants</B>\n",
    "\n",
    "inputs: bcf mpileup output<br>\n",
    "output: zipped vcf file <br>\n",
    "*notes: -mv (could use -m -v separately) m for calling snps and indels, v for only outputing variants from ref genome (don't need every position); --ploidy 1 specifies I am working with a haploid genome; -Oz specifies to zip the output sicne zipped vcf is required for vcf indexing and consensus commands "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd9a9974",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools call -mv --ploidy 1 -Oz -o strain.vcf.gz strain.mpileup.bcf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ab97ba2",
   "metadata": {},
   "source": [
    "<B>Step 5 - Index vcf</B>\n",
    "\n",
    "input: zipped vcf file<br>\n",
    "output: indexed vcf file (still zipped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93a925ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools index strain.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0131674e",
   "metadata": {},
   "source": [
    "<B>Step 6 - Filter vcf file for mapping quality and read depth</B>\n",
    "\n",
    "input: indexed zipped vcf file<br>\n",
    "output: filtered version of input file<br>\n",
    "*notes: can change threshold for mapping and depth, these seem reasonable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "becc4688",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools filter -e 'Qual<30 || DP<10' -Oz -o strain.filtered.vcf.gz strain.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bd87418",
   "metadata": {},
   "source": [
    "<B>Step 7 - Keep just snps and index again</B>\n",
    "\n",
    "input: filtered vcf file (zipped)<br>\n",
    "output: indexed vcf with only snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9339de7",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcftools view -v snps -Oz -o strain.snps.vcf.gz strain.filtered.vcf.gz\n",
    "\n",
    "bcftools index strain.snps.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b668f0f6",
   "metadata": {},
   "source": [
    "<B>Step 8 - Generate consensus sequence</B>\n",
    "\n",
    "inputs: reference fasta, indexed snps.vcf file<br>\n",
    "outputs: consensus sequence! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae74daf6",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "bcftools consensus -f ref.fasta strain.snps.vcf.gz > strain.consensus.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67cf1f41",
   "metadata": {},
   "source": [
    "<B>Step 9 - Change ID on fasta header line</B>\n",
    "\n",
    "This was neccessary since by default bcftools uses the ID on the reference header for the new consensus seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81fc6035",
   "metadata": {},
   "outputs": [],
   "source": [
    "sed -i '1s/.*/>new_id/' strain.consensus.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d88f7113",
   "metadata": {},
   "source": [
    "<H3>II. Generating consensus seqs from already assembled genomes</H3>\n",
    "\n",
    "*mostly same as when using sequencing reads, see above for specifics<br>\n",
    "*note - for same strains of same species I used -ax asm5, for different species I used asm10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61ea40a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 1 - map assembled genome (or part of it) to the reference\n",
    "minimap2 -ax asm5 ref.fasta strain.fasta > strain.sam\n",
    "\n",
    "# step 2 - convert sam to bam, sort, and index\n",
    "samtools view -bS strain.sam > strain.bam\n",
    "\n",
    "samtools sort strain.bam -o strain.sorted.bam\n",
    "\n",
    "samtools index strain.sorted.bam\n",
    "\n",
    "# step 3 - generate bcf pileup\n",
    "bcftools mpileup -f ref.fasta strain.sorted.bam -Ou -o strain.mpileup.bcf\n",
    "\n",
    "# step 4 - call variants\n",
    "bcftools call -mv --ploidy 1 -Oz -o strain.vcf.gz strain.mpileup.bcf\n",
    "\n",
    "# step 5 - index vcf file\n",
    "bcftools index strain.vcf.gz\n",
    "\n",
    "# step 6 - filter and re-index vcf \n",
    "bcftools filter -e 'QUAL<30' -Oz -o strain.filtered.vcf.gz strain.vcf.gz\n",
    "\n",
    "bcftools index strain.filtered.vcf.gz\n",
    "\n",
    "#step 7 - get snps only and re-index\n",
    "bcftools view -v snps -Oz -o strain.snps.vcf.gz strain.filtered.vcf.gz\n",
    "\n",
    "bcftools index strain.snps.vcf.gz\n",
    "\n",
    "# step 8 - generate consensus sequence\n",
    "bcftools consensus -f ref.fasta strain.snps.vcf.gz > strain.consensus.fasta\n",
    "\n",
    "# step 9 - rename fasta header id\n",
    "sed -i '1s/.*/>new_id/' strain.consensus.fasta"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
